File: 2021/archived/Day17/Day17.scala
import scala.io.Source
import scala.annotation.tailrec

@main def day17 =
    val abs = Math.abs
    val max = Math.max

    def boundX(x1: String, x2: String) =
        (abs(x1.toInt) :: abs(x2.toInt) :: Nil).sorted

    def boundY(y1: String, y2: String) =
        (y1.toInt :: y2.toInt :: Nil).sorted

    def inZone(minX: Int, maxX: Int, minY: Int, maxY: Int)(x: Int, y: Int) =
        minX <= x && x <= maxX && minY <= y && y <= maxY

    def shoot(maxX: Int, minY: Int, vx: Int, vy: Int) =
        @tailrec
        def loop(vx: Int, vy: Int, x: Int, y: Int, h: Int): (Int, Int, Int) =
            if x > maxX || y < minY then (x - vx, y - vy, h)
            else
                val nvx = max(vx - 1, 0)
                val nvy = vy - 1
                loop(nvx, nvy, x + nvx, y + nvy, max(h, y + nvy))

        loop(vx, vy, vx, vy, Int.MinValue)

    def multiShoot(minX: Int, maxX: Int, minY: Int, maxY: Int) =
        (for
            vx <- maxX to 0 by -1
            vy <- minY to abs(minY) - 1
        yield (vx, vy))
            .map({ case (vx, vy) =>
                shoot(maxX, minY, vx, vy)
            })
            .filter({ case (x, y, _) =>
                inZone(minX, maxX, minY, maxY)(x, y)
            })
            .toVector

    def firstPart(landing: Vector[(Int, Int, Int)]) =
        landing.maxBy(_._3)._3

    def secondPart(landing: Vector[(Int, Int, Int)]) = landing.size

    val raw = Source.fromFile("input.txt").getLines.next

    val (minX :: maxX :: Nil, minY :: maxY :: Nil) =
        "([0-9\\-]+)".r.findAllIn(raw).map(_.strip).toList match
            case x1 :: x2 :: y1 :: y2 :: Nil => (boundX(x1, x2), boundY(y1, y2))
            case _ => sys.error(s"Invalid input ${raw}")

    val landing = multiShoot(minX, maxX, minY, maxY)

    println(s"First part: ${firstPart(landing)}")
    println(s"Second part: ${secondPart(landing)}")

File: 2021/archived/Day17/Day17.fs
module Day17

open System
open System.IO
open System.Text.RegularExpressions

let (<.>) f x y = f y x

let boundX x1 x2 =
    let l =
        [ x1 |> int |> abs; x2 |> int |> abs ]
        |> List.sort

    (l.Head, l.Tail.Head)

let boundY y1 y2 =
    let l = [ y1 |> int; y2 |> int ] |> List.sort
    (l.Head, l.Tail.Head)

let inZone minX maxX minY maxY x y =
    minX <= x && x <= maxX && minY <= y && y <= maxY

let shoot maxX minY vx vy =

    let rec loop vx vy x y h =
        if x > maxX || y < minY then
            x - vx, y - vy, h
        else
            let vx = max (vx - 1) 0
            let vy = vy - 1
            loop vx vy (x + vx) (y + vy) (max h (y + vy))

    loop vx vy vx vy Int32.MinValue

let multiShoot minX maxX minY maxY =
    let inZone = inZone minX maxX minY maxY

    [ for vx in maxX .. -1 .. 0 do
          for vy in minY .. (minY |> abs |> ((-) <.> 1)) -> vx, vy ]
    |> List.map (fun (vx, vy) -> shoot maxX minY vx vy)
    |> List.filter (fun (x, y, _) -> inZone x y)

let firstPart landing =
    landing
    |> Seq.maxBy (fun (_, _, h) -> h)
    |||> (fun _ _ h -> h)

let secondPart landing = landing |> Seq.length

let raw = File.ReadAllLines "input.txt" |> Array.head

let (minX, maxX), (minY, maxY) =
    match Regex.Matches(raw, "([0-9\-]+)")
          |> Seq.cast<Match>
          |> Seq.toList
        with
    | [ x1; x2; y1; y2 ] -> (boundX x1.Value x2.Value), (boundY y1.Value y2.Value)
    | _ -> failwithf "Invalid input %s" raw

let landing = multiShoot minX maxX minY maxY

firstPart landing |> printfn "First part: %d"
secondPart landing |> printfn "Second part: %d"
